wireguard-rs/src/run/runner.rs

```diff
@@ -142,37 +158,39 @@ fn spawn_tun_event_loop<'scope, 'env, T: Tun, B: PlatformUDP, S: Status>(
                     exit(ExitCode::TUNDeviceError as i32);
                 }
                 Ok(TunEvent::Up(mtu)) => {
-                    log::info!("Tun up (mtu = {})", mtu);
-                    let _ = wireguard_config.up(mtu); // TODO: handle
+                    config_sender
+                        .send(ConfigMessage::TunUp(mtu))
+                        .expect("channel is open while this loop is running");
                 }
                 Ok(TunEvent::Down) => {
-                    log::info!("Tun down");
-                    wireguard_config.down();
+                    config_sender
+                        .send(ConfigMessage::TunDown)
+                        .expect("channel is open while this loop is running");
                 }
             }
         }
     })
 }
 
-fn spawn_uapi_server<'scope, 'env, T: Tun, B: PlatformUDP, U: BindUAPI + Send + Sync>(
+fn spawn_uapi_server<'scope, 'env>(
     thread_scope: &'scope thread::Scope<'scope, 'env>,
-    wireguard_config: &'env WireGuardConfig<T, B>,
-    uapi: &'env U,
     tun_reader_jobs_running: &'env AtomicBool,
-) -> ScopedJoinHandle<'scope, ()>
-where
-    <U as BindUAPI>::Stream: Send,
-    <U as BindUAPI>::Stream: 'env,
-{
+    uapi: &'env <plt::UAPI as PlatformUAPI>::Bind,
+    config_sender: mpsc::Sender<ConfigMessage>,
+) -> ScopedJoinHandle<'scope, ()> {
     thread_scope.spawn(|| {
+        let config_sender = config_sender;
+
         while tun_reader_jobs_running.load(Ordering::Acquire) {
             // accept and handle UAPI config connections
             match uapi.connect() {
                 Ok(stream) => {
-                    thread_scope.spawn(|| {
-                        let mut stream = stream;
-                        uapi::handle(&mut stream, wireguard_config);
-                    });
+                    spawn_uapi_config_message_handler(
+                        thread_scope,
+                        tun_reader_jobs_running,
+                        config_sender.clone(),
+                        stream,
+                    );
```

Comment:
Not sure I think this is simpler.

It makes it harder to test `uapi::handle` in isolation: the method needs to take `thread_scope` and `tun_reader_jobs_running`